xtm/xtm_api.cc

```diff
@@ -0,0 +1,353 @@
+	    fcntl(queue->consumer_write_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    (is_producer_need_notifications &&
+	    (fcntl(queue->producer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	     fcntl(queue->producer_write_fd, F_SETFL, O_NONBLOCK) < 0))) {
+		save_errno = errno;
+		goto close_producer_fd;
+	}
+	if (queue->queue.create(size) < 0) {
+		save_errno = EINVAL;
+		goto close_producer_fd;
+	}
+	return queue;
+
+close_producer_fd:
+	if (is_producer_need_notifications) {
+		close(queue->producer_read_fd);
+		if (queue->producer_read_fd != queue->producer_write_fd)
+			close(queue->producer_write_fd);
+	}
+close_consumer_fd:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (queue->is_need_to_close_readfd &&
+	    queue->is_producer_need_notifications &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (queue->is_producer_need_notifications &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (queue->is_need_to_close_readfd &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) != 1) {
+		errno = ENOBUFS;
+		if (queue->is_producer_need_notifications) {
+			__atomic_store_n(&queue->is_notification_expected,
+					 true, __ATOMIC_RELEASE);
+		}
+		return -1;
+	}
+	return 0;
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+int
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.end();
+	if (queue->is_producer_need_notifications &&
+	    __atomic_load_n(&queue->is_notification_expected,
+			    __ATOMIC_ACQUIRE)) {
+		__atomic_store_n(&queue->is_notification_expected,
+				 false, __ATOMIC_RELEASE);
+		if (notify_fd(queue->producer_write_fd) < 0)
+			return -1;
```

Comment:
This way we lose "cnt" value.
I suggest returning "-(cnt + 1)" here. This way it is easy to check for error ("< 0") and calculate cnt if someone really needs it. No extra memory writes, just registers.

Anyone have better ideas?

Reply:
@Gerold103 What do you think about it?

Reply:
Sounds good to me.